src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.Tasks/JoinVerticals.cs

```diff
@@ -0,0 +1,172 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using Microsoft.Build.Utilities;
+using Microsoft.Build.Framework;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.DotNet.UnifiedBuild.Tasks;
+
+public class JoinVerticals : Microsoft.Build.Utilities.Task
+{
+    /// <summary>
+    /// Paths to Verticals Manifests
+    /// </summary>
+    [Required]
+    public required ITaskItem[] VerticalManifest { get; init; }
+
+    /// <summary>
+    /// Optional subset of the Verticals we want to join, separated by ;
+    /// </summary>
+    /// <example>Windows_x64;Windows_x86</example>
+    public string? VerticalSubSet { get; set; }
+
+    /// <summary>
+    /// Name of the main vertical that we'll take all artifacts from
+    /// </summary>
+    [Required]
+    public required string MainVertical { get; init; }
+
+    /// <summary>
+    /// Azure DevOps build id
+    /// </summary>
+    [Required]
+    public required string BuildId { get; init; }
+
+    /// <summary>
+    /// Azure DevOps token, required scopes: "Build (read)"
+    /// </summary>
+    [Required]
+    public required string AzureDevOpsToken { get; init; }
+
+    /// <summary>
+    /// Azure DevOps organization
+    /// </summary>
+    [Required]
+    public required string AzureDevOpsOrg { get; init; }
+
+    /// <summary>
+    /// Azure DevOps project
+    /// </summary>
+    [Required]
+    public required string AzureDevOpsProject { get; init; }
+
+    /// <summary>
+    /// Temporary folder where artifacts will be downloaded and extracted
+    /// </summary>
+    [Required]
+    public required string TmpFolder { get; init; }
+
+    /// <summary>
+    /// Folder where packages and assets will be stored
+    /// </summary>
+    [Required]
+    public required string OutputFolder { get; init; }
+
+    /// <summary>
+    /// If true, the task will copy all packages to the same output folder, they won'y be separated by shipping/nonshipping or repo.
+    /// Set to true for the final join, false for all other joins.
+    /// </summary>
+    public required bool FlatCopy { get; init; }
+
+    private readonly string _packageElementName = "Package";
+    private readonly string _blobElementName = "Blob";
+    private readonly string _idAttribute = "Id";
+    private readonly string _verticalNameAttribute = "VerticalName";
+    private readonly string _artifactNameSuffix = "_Artifacts";
```

Comment:
Maybe if these were `consts` we could make some of the methods below static

Reply:
Originally,  tried doing that, but then we'd have to pass the logger as a parameter too

Reply:
It's still better to have these as consts though? Or what is the thinking behind having fields?

I believe at least one method did not even use logger that I noticed.

Reply:
made these const, and made static what could be static